### ISSUE TITLE
Do not watch ignored files

### ISSUE THREAD

**Issue by FDiskas** [withastro/astro#8224]
### Astro info

```block
Astro version            v2.10.11
Package manager          npm
Platform                 darwin
Architecture             arm64
Adapter                  Couldn't determine.
Integrations             @astrojs/tailwind, relative-links
```

### What browser are you using?

chrome

### Describe the Bug

![image](https://github.com/withastro/astro/assets/468006/e83825b8-077d-46d0-8b17-c667f17b72a7)

### What's the expected result?

It should not reload the project if some ignored file is updated.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-jqwvp1?file=.DS_Store&on=stackblitz

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by FDiskas** [comment]
I believe that astro could use gitignore file or have his own

**Comment by natemoo-re** [comment]
Hmm, yeah we might need to read the user's `.gitignore` file and then filter out updates. However, things like `node_modules` are traditionally ignored by git but should still affect HMR, so I don't know what the right answer is here.

**Comment by cornedor** [comment]
I have another use case. I'm using ReScript to compile some scripts. When ReScript's watcher is running, it will create a `.bsb.lock` file that is quickly removed. Sometimes so quickly that the watcher will pick it up, but the file will be removed by the time it is handled, causing the Astro watcher to crash.

**Comment by ematipico** [comment]
The `.gitignore` support should be opt-in to avoid unpleasant behaviours.

**Comment by cornedor** [comment]
Perhaps for some cases, it is better to add a configuration option to disable watcher for a list (glob?) of files?

**Comment by ematipico** [comment]
> Perhaps for some cases, it is better to add a configuration option to disable watcher for a list (glob?) of files?

You can customise `vite` behaviour directly form your `astro.config.mjs`

```js
export default defineConfig({
	vite: {
		server: {
			watch: {
				
			}
		}
	}
})
```

Here's the options: https://vitejs.dev/config/server-options.html#server-watch

**Comment by rishi-raj-jain** [comment]
@natemoo-re Per the solution suggested by @ematipico, I think we don't need a resolution for this. IMO it should be done in vite as in the comment above.

**Comment by bluwy** [comment]
The issue here is that Vite and Astro logs their HMR update logs differently. Vite only logs when an [actual HMR update happen](https://github.com/vitejs/vite/blob/d3b96e03083ba3a9d19eca09bd2853094775a5a8/packages/vite/src/node/server/hmr.ts#L205-L209). Astro suppresses it https://github.com/withastro/astro/blob/fca6892f8d6a30ceb1e04213be2414dd4cb4d389/packages/astro/src/core/create-vite.ts#L246 (Vite's HMR updates are `info`, but Astro only logs warnings/errors from Vite) and logs whenever [any update happen](https://github.com/withastro/astro/blob/fca6892f8d6a30ceb1e04213be2414dd4cb4d389/packages/astro/src/vite-plugin-astro/hmr.ts#L125-L130).

Technically Vite is watching the entire `root` that this is happening. But it cannot ignore watching things in `.gitignore` because there can be ephemeral and dynamically generated file used there. So the solution here perhaps is to see if Astro could re-use Vite's logger, instead of re-implementing it.

**Comment by natemoo-re** [comment]
> Technically Vite is watching the entire `root` that this is happening. But it cannot ignore watching things in `.gitignore` because there can be ephemeral and dynamically generated file used in your source code. So the solution here perhaps is to see if Astro could re-use Vite's logger, instead of re-implementing it.

Yeah @bluwy is totally correct here. There are situations where you may want to `.gitignore` some files but they should still be treated as input modules that need to be watched for updates. So I don't believe Astro should be blindly reading a user's `.gitignore` file.

The best solution would be to align Astro's custom handling with Vite's implementation. Seems like we might be able to use the Vite dev server's `ws` channel for this. https://github.com/vitejs/vite/blob/d3b96e03083ba3a9d19eca09bd2853094775a5a8/packages/vite/src/node/server/hmr.ts#L210C1-L213C5

**Comment by cdtut** [comment]
`outDir` should be ignored by default because never a reason to watch dist folder. When `astro build` is run when `astro dev` is already running there are many output lines such as `[astro] update /dist/server/entry.mjs` which should not be.

**Comment by cdtut** [comment]
@bluwy Can `outDir` be ignored by default for the PR?
